pkg/handlers/httpproxy/inject.go

```diff
@@ -122,3 +137,78 @@ func (h *httpProxyPatchHandler) GeneratePatches(
 		},
 	)
 }
+
+func (h *httpProxyPatchHandler) detectNoProxy(
+	ctx context.Context,
+	req *runtimehooksv1.GeneratePatchesRequest,
+) ([]string, error) {
+	clusterKey := types.NamespacedName{}
+
+	for _, item := range req.Items {
+		if item.HolderReference.Kind == "Cluster" &&
+			item.HolderReference.APIVersion == capiv1.GroupVersion.String() {
+			clusterKey.Name = item.HolderReference.Name
+			clusterKey.Namespace = item.HolderReference.Namespace
+		}
+	}
+
+	if clusterKey.Name == "" {
+		return nil, errors.New("failed to detect cluster name from GeneratePatch request")
+	}
+
+	cluster := &capiv1.Cluster{}
+	if err := h.client.Get(ctx, clusterKey, cluster); err != nil {
+		return nil, err
+	}
+
+	return generateNoProxy(cluster), nil
+}
+
+// generateNoProxy creates default NO_PROXY values that should be applied on cluster
+// in any environment and are preventing the use of proxy for cluster internal
+// networking.
+func generateNoProxy(cluster *capiv1.Cluster) []string {
+	noProxy := []string{
+		"localhost",
+		"127.0.0.1",
+	}
+
+	if cluster.Spec.ClusterNetwork != nil &&
+		cluster.Spec.ClusterNetwork.Pods != nil {
+		noProxy = append(noProxy, cluster.Spec.ClusterNetwork.Pods.CIDRBlocks...)
+	}
+
+	if cluster.Spec.ClusterNetwork != nil &&
+		cluster.Spec.ClusterNetwork.Services != nil {
+		noProxy = append(noProxy, cluster.Spec.ClusterNetwork.Services.CIDRBlocks...)
+	}
+
+	serviceDomain := "cluster.local"
+	if cluster.Spec.ClusterNetwork != nil &&
+		cluster.Spec.ClusterNetwork.ServiceDomain != "" {
+		serviceDomain = cluster.Spec.ClusterNetwork.ServiceDomain
+	}
+
+	noProxy = append(noProxy, []string{
+		"kubernetes",
+		"kubernetes.default",
+		".svc",
+	}...)
+
+	noProxy = append(noProxy, mapServiceDomain(serviceDomain, ".svc")...)
+
+	return noProxy
+}
+
+// mapServiceDomain generates NO_PROXY values for service domain to allow use
+// of short service name in cluster.
+// Example: my.custom => [.svc, .svc.my, .svc.my.custom].
+func mapServiceDomain(serviceDomain, svcName string) []string {
+	mapped := []string{svcName}
+	progress := svcName
+	for _, part := range strings.Split(serviceDomain, ".") {
```

Comment:
Will this always be in `*.*` format?

Reply:
I think it could be an arbitrary number of segments but would need to check. I also think we could probably get away just with adding the whole cluster domain rather than individual parts

Reply:
Validation in kubeadm at https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/apis/kubeadm/validation/validation.go#L531 defined in regexp https://github.com/kubernetes/apimachinery/blob/master/pkg/util/validation/validation.go#L205

Reply:
We should add pattern validation to the open API schema, I need to do the same for cert sans

Reply:
Thanks for confirming this @jimmidyson. The parser implemented here will map any number of parts so I think it works correctly.

Reply:
I'm not sure we need to add every part of the cluster domain though. I would expect only to include the full domain, otherwise this could have some  (highly unlikely but possible) invalid behaviour if there is an valid external domain that is a prefix of the cluster domain.

`my.custom => [.svc, svc.my.custom]`

Reply:
Changed the behavior to only include values as described.